xarray/backends/core.py

```diff
@@ -0,0 +1,425 @@
+        array to a tuple of slice objects, e.g., ``x[key] = value``, where
+        ``key`` has type ``Tuple[slice, ...]`` and length equal to the
+        dimensionality of the array, and ``value`` is a ``numpy.ndarray``.
+        """
+        raise NotImplementedError
+
+    def get_writable_array(self, name):
+        # type: (Hashable,) -> WritableDuckArray
+        """Return a writable array corresponding to an existing variable.
+
+        This method is only needed if you want the DataStore to support partial
+        writes, e.g., appending to existing variables.
+
+        Parameters
+        ----------
+        name : Hashable
+            Variable name. How to handle non-existing names is up to the
+            DataStore class. However, xarray will never call this method unless
+            a variable with the given name has already been verified to exist
+            as a member of the mapping returned by `get_variables()`.
+
+        Returns
+        -------
+        Writable array-like object, see `create_variable` for details.
+        """
+        # Note: this mostly exists for the benefit of future support for partial
+        # reads -- we don't actually make use of this in the current version of
+        # xarray.
+        raise NotImplementedError
+
+    def set_attribute(self, name, value):
+        # type: (Hashable, Any) -> None
+        """Set a global attribute on the DataStore."""
+        raise NotImplementedError
+
+    def sync(self):
+        """Synchronize writes to this DataStore."""
+        pass
+
+    def get_write_lock(self, name, region=Ellipsis):
+        # type: (Hashable, Union[Ellipsis, Tuple[slice, ...]]) -> object
+        """Return a lock for writing a given variable.
+
+        This method may be useful for DataStores that from which data is
+        written in parallel (e.g., with dask).
+
+        Parameters
+        ----------
+        name : Hashable
+            Variable name.
+        region : Union[Ellipsis, Tuple[slice, ...]], optional
+            Region in the variable, e.g., valid key argument to `data[key]`,
+            for which to get a lock.
+
+        Returns
+        -------
+        threading.Lock() ducktype (that is, an object with `acquire` and
+        `release` methods), or None, if no lock is necessary.
+        """
+        # Again, we actually have a use for the region argument? Could be useful
+        # to ensure writes to zarr are safe.
+        return None
+
+
+class InMemoryDataStore(AbstractWritableDataStore):
+    """Stores variables and attributes directly in OrderedDicts.
+
+    This store exists internal testing purposes, e.g., for integration tests
+    with dask.array that will not need to write actual data to disk.
+    """
+    def __init__(self):
+        self._variables = OrderedDict()
+        self._attributes = OrderedDict()
+        # do we need locks? are writes to NumPy arrays thread-safe?
```

Comment:
I think the answer is yes, we need locks (no writes are not thread safe), to this question but I imagine @mrocklin can give the final word.

Reply:
We never do overlapping writes though, right?  I've found that locks are not necessary as long as the underlying data store's chunking doesn't overlap poorly with how we're writing chunks.  Given that NumPy arrays are entirely fine-grained this doesn't seem like it would be an issue.